docs/design/tools/illink/feature-switch-attributes.md

```diff
@@ -0,0 +1,482 @@
+
+- The "feature" of individual assembly files being available in an app is represented as:
+  - `RequiresAssemblyFilesAttribute`
+
+Similarly, not all features come with an attribute. Some features don't tie into functionality which is designed to produce warnings if called. For example:
+
+- The "feature" of verifying that open generics are dynamically instantiated with trim-compatible arguments is represented as:
+  - `VerifyOpenGenericServiceTrimmability` property
+  - `"Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability"`
+
+- The "feature" of supporting or not supporting globalization is represented as:
+  - `GlobalizationMode.Invariant` property
+  - `"System.Globalization.Invariant"`
+
+No warnings are produced just because these features are enabled/disabled. Instead, they are designed to be used to change the behavior, and possibly remove unneccessary code when publishing. It's easy to imagine adding an attribute like `RequiresVariantGlobalization` that would be used to annotate APIs that rely on globalization support, such that analysis warnings would prevent accidentally pulling in the globalization stack from code that is supposed to be using invariant globalization. We don't want to do this for every feature; typically we only do this for features that represent large cross-cutting concerns, and are not available with certain app models (trimming, AOT compilation, single-file publishing).
+
+However, any attribute-based model that we pick to represent features should be able to tie in with all three representations.
+
+### Unified attribute model for feature switches and guards
+
+We would like for the attribute model to have a consistent way to refer to a feature in `FeatureGuardAttribute` or `FeatureSwitchAttribute`. The proposal is to support an attribute model for features that have an associated `Requires` attribute, and use that attribute type uniformly to refer to the feature from `FeatureGuardAttribute` and `FeatureSwitchAttribute`.
+
+For example, the feature switch for "dynamic code support" might look like this:
+```csharp
+public class RuntimeFeature
+{
+    [FeatureSwitch<RequiresDynamicCode>]
+    public static bool IsDynamicCodeSupported => // ...
+}
+
+class RequiresDynamicCodeAttribute : Attribute, IFeatureAttribute<RequiresDynamicCodeAttribute> {
+    public static string FeatureName => "System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported";
+}
+```
+and a feature guard for "dynamic code compilation" might look like this:
+```csharp
+public class RuntimeFeature
+{
+    [FeatureGuard<RequiresDynamicCode>]
+    public static bool IsDynamicCodeSupported => // ...
+}
+```
+
+The attribute definitions to support this might look like this:
+```csharp
+class FeatureGuardAttribute<T> : Attribute
+  where T : Attribute, IFeatureAttribute<T> {}
+
+interface IFeatureAttribute<TSelf> where TSelf : Attribute {
+    static abstract string FeatureName { get; }
+}
```

Comment:
Could this be a custom attribute instead of an interface?

1. For non-execution scenarios it's an annoyance having to resolve virtual calls. The interface method could be implemented on a base, or by a default interface method and the analysis tool needs to deal with all of that just to get a string out. Custom attribute has it in the blob.
2. Static virtuals and generic attributes will not compile on netstandard2.0. Might not be a concern for trimming/AOT, but could be a concern for some other feature.

Reply:
+1 on this - as much as I would love to be able to use all the cool new things, the inability to compile this for `netstandard2.0` is pretty bad.
Using an attribute would effectively provide the same information. We would have to enforce its presence via analysis though.

Reply:
The point are libraries which multi-target (our recommended approach). If such a library introduces its own feature guard/switch, it would need to use this. Some authors don't like ifdefs, so they may want to backfill the attributes, which would then not work for netstandard2.0. It's probably not very common, but still worth consideration.

The robustness and ease of the tooling is a bigger argument for me, and since this is not going to be common, we don't need the API to be super clean and super easy to use.

I don't feel strongly about this. I do agree with Michal's concerns though.

Reply:
The interface approach introduces other class of robustness issues - there's nothing to enforce that the implementation of the method is a string literal, so we'd either need another analyzer for this to ensure string literal, or live with the fact that it might not be a string literal and have a way to report that.

Other thing is that the Roslyn analyzer wouldn't have visibility into the method body if it's in a different compilation unit.